lib/rex/post/meterpreter/ui/console/command_dispatcher/bofloader.rb

```diff
@@ -0,0 +1,300 @@
+            end
+
+            if compile
+              bof_data = compile_c(bof_filename)
+              return unless bof_data
+            else
+              bof_data = ::File.binread(bof_filename)
+            end
+
+            # loading all data will hang on invalid files like DLLs, so only parse the 20-byte header at first
+            parsed = Metasm::COFF.decode_header(bof_data[0...20])
+            bof_arch = { # map of metasm to metasploit architectures
+              'AMD64' => ARCH_X64,
+              'I386' => ARCH_X86
+            }.fetch(parsed.header.machine, nil)
+
+            unless bof_arch
+              print_error('Unable to determine the file architecture.')
+              return
+            end
+            unless bof_arch == client.arch
+              print_error("The file architecture is incompatible with the current session (file: #{bof_arch} session: #{client.arch})")
+              return
+            end
+
+            parsed = Metasm::COFF.decode(bof_data)
+            unless (executable_symbols = get_executable_symbols(parsed)).include?(entry)
+              print_error("The specified entry point was not found: #{entry}")
+              print_error("Available symbols: #{executable_symbols.join(', ')}")
+              return
+            end
+
+            begin
+              output = client.bofloader.execute(bof_data, args_format: bof_args_format, args: bof_args, entry: entry)
+            rescue Rex::Post::Meterpreter::Extensions::Bofloader::BofPackingError => e
+              print_error("Error processing the specified arguments: #{e.message}")
+              return
+            end
+
+            if output.nil?
+              print_status('No output returned from bof')
+            else
+              print_line(output)
+            end
+          end
+
+          private
+
+          def compile_c(source)
+            if client.arch == ARCH_X86
+              mingw = Metasploit::Framework::Compiler::Mingw::X86.new
+            elsif client.arch == ARCH_X64
+              mingw = Metasploit::Framework::Compiler::Mingw::X64.new
+            else
+              print_error("Unsupported client architecture: #{client.arch}")
+              return
+            end
+
+            unless mingw.class.available?
+              print_error("#{mingw.mingw_bin} is unavailable, can not compile source code")
+              return
+            end
+
+            ::Dir::Tmpname.create([::File.basename(source, '.c'), '.o']) do |destination|
+              output, status = Open3.capture2e(mingw.mingw_bin, '-c', source, '-I', Metasploit::Framework::Compiler::Mingw::INCLUDE_DIR, '-o', destination)
+              unless status.exitstatus == 0
+                print_error("Compilation exited with error code: #{status.exitstatus}")
+                print_line(output) unless output.blank?
+                return
+              end
+
+              bof_data = ::File.binread(destination)
+              ::File.delete(destination)
```

Comment:
Is this needed? I think cleanup should happen automagically?

https://github.com/ruby/ruby/blob/cc7f6fe73454743da87d73ddb26626503fae248f/lib/tmpdir.rb#L128-L152

Reply:
Fixed in e862141eaa25d933448337838e38b03adbc1838b.